noir-r1cs/src/noir_proof_scheme.rs

```diff
@@ -94,18 +100,54 @@ impl NoirProofScheme {
         (self.r1cs.num_constraints(), self.r1cs.num_witnesses())
     }
 
+    pub fn read_witness(&self, prover_toml: impl AsRef<Path>) -> Result<InputMap> {
+        let (input_map, _expected_return) =
+            read_inputs_from_file(prover_toml.as_ref(), self.witness_generator.abi())?;
+
+        Ok(input_map)
+    }
+
+    pub fn generate_witness(&self, input_map: &InputMap) -> Result<WitnessMap<NoirFieldElement>> {
+        let span = span!(Level::INFO, "generate_witness").entered();
+        let solver = Bn254BlackBoxSolver::default();
+        let mut output_buffer = Vec::new();
+        let mut foreign_call_executor = DefaultForeignCallBuilder {
+            output:       &mut output_buffer,
+            enable_mocks: false,
+            resolver_url: None,
+            root_path:    None,
+            package_name: None,
+        }
+        .build();
+
+        let initial_witness = self.witness_generator.abi().encode(input_map, None)?;
+
+        let mut witness_stack = nargo::ops::execute_program(
+            &self.program,
+            initial_witness,
+            &solver,
+            &mut foreign_call_executor,
+        )?;
+
+        drop(span);
```

Comment:
Redundant, it would drop around here anyway as the scope ends (unless you specifically want to exclude these last couple operations like `.pop()`)